### PR TITLE
Loop module directories first

### PR DESCRIPTION
**Pull request by unmultimedio**
We can safely start looping module directories, and for each one we can traverse all branches, instead of keeping a state of different expected sync points and start/stop sync points for each module/branch combination. It makes the code more readable, and it's still safe to traverse the Git tree once per module directory (as they should have different module identity targets).

### CODE REVIEW

File: .github/workflows/buf-shadow-sync.yaml [hunk 1]
```diff
@@ -1,7 +1,8 @@
 name: Buf shadow sync
 
 on:
-  push: # Any push, regardless if it contains proto changes or not
+  # TODO: reenable after multibranch improvements
+  # push: # Any push, regardless if it contains proto changes or not
```
Comment (unmultimedio) [inline .github/workflows/buf-shadow-sync.yaml:5]: https://github.com/bufbuild/buf/pull/2409

File: private/buf/bufsync/lookback_start_point_test.go [hunk 2]
```diff
@@ -1,214 +0,0 @@
...
-					modulesToSyncInCommit := make(map[string]*bufmodulebuild.BuiltModule)
-					for moduleDir := range commitModulesToSync.modules {
-						modulesToSyncInCommit[moduleDir] = nil // no need for a built module in this test
-					}
-					syncableCommits = append(syncableCommits, &syncableCommit{
-						commit:  &mockCommit{commitModulesToSync.commitHash},
-						modules: modulesToSyncInCommit,
-					})
-				}
-				gotStartPoints, err := s.lookbackStartingPoints(context.Background(), branchName, syncableCommits)
```
Comment (unmultimedio) [inline private/buf/bufsync/lookback_start_point_test.go:158]: This func was not needed anymore, so its test.

File: private/buf/bufsync/syncer.go [hunk 3]
```diff
@@ -57,13 +56,15 @@ type syncer struct {
...
 	commitsToTags map[string][]string
-	// branchesToModulesForSync holds all the branches, module directories, and module identity
-	// targets for those directories, prepared before syncing either from its identity override or
-	// HEAD commit. (branch:moduleDir:targetModuleIdentity)
-	branchesToModulesForSync map[string]map[string]bufmoduleref.ModuleIdentity
-	// modulesToBranchesLastSyncPoints holds remote expected sync points for module identity and its
+	// modulesDirsToBranchesToIdentities holds all the module directories, branches, and module identity
+	// targets for those directories+branches, prepared before syncing either from its identity
+	// override or HEAD commit. (moduleDir:branch:targetModuleIdentity)
+	modulesDirsToBranchesToIdentities map[string]map[string]bufmoduleref.ModuleIdentity
```
Comment (unmultimedio) [inline private/buf/bufsync/syncer.go:62]: This represents the main change for this PR, no longer `branch -> moduleDir` but `moduleDir -> branch`

File: private/buf/bufsync/syncer.go [hunk 4]
```diff
@@ -280,217 +309,177 @@ func (s *syncer) validateDefaultBranch(ctx context.Context, moduleIdentity bufmo
...
-// error handler, or are a found sync point.
-func (s *syncer) branchSyncableCommits(ctx context.Context, branch string) ([]*syncableCommit, error) {
-	branchModulesForSync, ok := s.branchesToModulesForSync[branch]
-	if !ok || len(branchModulesForSync) == 0 {
-		// branch should not be synced, or no modules to sync in that branch
-		return nil, nil
-	}
-	// Copy all branch modules to sync and mark them as pending, until its starting sync point is
-	// reached. They'll be removed from this list as its initial sync point is found.
-	pendingModules := s.copyBranchModulesSync(branch, branchModulesForSync)
```
Comment (unmultimedio) [inline private/buf/bufsync/syncer.go:391]: No need for the concept of "pending modules" anymore, this func is invoked once per module+branch, so each time it's just a single module.

File: private/buf/bufsync/syncer.go [hunk 5]
```diff
@@ -666,140 +605,70 @@ func (s *syncer) readModuleAt(
...
-		// branch should not be synced, or no modules to sync in that branch
-		return nil
-	}
-	moduleToStartPoint, err := s.lookbackStartingPoints(ctx, branch, syncableCommits)
-	if err != nil {
-		return fmt.Errorf("find lookback starting points for branch %s: %w", branch, err)
-	}
-	timeLimit := clock.Now().Add(-lookbackTimeLimit)
-	stopLoopErr := errors.New("stop loop")
-	for moduleDir, startPoint := range moduleToStartPoint {
```
Comment (unmultimedio) [inline private/buf/bufsync/syncer.go:684]: No need for this outer loop. The `backfillTags` func is not invoked once per module directory.

File: private/buf/bufsync/syncer.go [hunk 6]
```diff
@@ -818,44 +687,15 @@ func (s *syncer) printSyncPreparation() {
...
+		zap.Any("branchesModulesToSync", s.modulesDirsToBranchesToIdentities),
+		zap.Any("modulesBranchesSyncPoints", s.modulesToBranchesExpectedSyncPoints),
 	)
 }
 
-// syncableCommit holds the modules that need to be synced in a git commit.
+// syncableCommit holds the built module that need to be synced in a git commit.
 type syncableCommit struct {
-	commit  git.Commit
-	modules map[string]*bufmodulebuild.BuiltModule // moduleDir:builtModule
```
Comment (unmultimedio) [inline private/buf/bufsync/syncer.go:848]: A single syncable commit does not contain an array of built modules, but just a single built module (the module dir is implied).

File: private/buf/bufsync/commits_to_sync_test.go [hunk 7]
```diff
@@ -130,6 +125,7 @@ func TestCommitsToSyncWithNoPreviousSyncPoints(t *testing.T) {
 type mockErrorHandler struct{}
 
 func (*mockErrorHandler) HandleReadModuleError(readErr *ReadModuleError) LookbackDecisionCode {
+	// fmt.Printf("read error: %v\n", readErr)
```
Comment (seankimdev) [inline private/buf/bufsync/commits_to_sync_test.go:128]: Is this comment intentionally left here?

Comment (unmultimedio) [reply private/buf/bufsync/commits_to_sync_test.go:128]: Yes, same as L107, useful sometimes when debugging the test. I think I'll just remove them.

Comment (unmultimedio) [reply private/buf/bufsync/commits_to_sync_test.go]: Done

**Review by doriable — APPROVED** [review]

**Pull request merged**